pkg/river/token/builder/blocks_nesting_test.go

```diff
@@ -0,0 +1,694 @@
+package builder_test
```

Comment:
Unfortunately, I find this file fairly hard to read:

* The names for identifiers are quite long and hard to parse, such as `OutMatchDefInPtrBlkOptMatchDef`. 
* I find the test harness code here hard to grok beacuse of the coupling between test types (i.e., types that will be used for encoding/decoding in tests) and the tests to perform (i.e., the implementation of the `testFactory` method). I would prefer a giant test table in a single test function over what is being done now. 

Can we find a way to condense this file down to make it easier to scan and identify what's being tested? Also, if possible, can we try to build on top of tests to get test coverage rather than trying to test every combination of scenarios? (For example, I am not sure it's necessary for all the pointer tests to exist, since they don't behave any differently than non-pointer examples, minus what the definition of the "zero value" is).

Reply:
Yeah agreed, it's been hard to name these tests and keep them easy to read, because structs need to have methods so they are declared outside of the test function. I've attempted to at least keep the test cases next to the structs definitions.

What do you propose I call a test case structure that tests `outer with zero value default, inner struct block, optional with a default value`? I could just call them `TestStructN` and keep more descriptive test names. I'm not sure it is going to help though. 

I do think we should have more edge case tests and I wanted to cover a large number of cases, because I found it's easy to get surprising behaviour and pretty time consuming to figure it out in the code.

Reply:
> Also, if possible, can we try to build on top of tests to get test coverage rather than trying to test every combination of scenarios?

I wouldn't rely on coverage too much here, because we use reflection a lot and for example executing an if statement on `v.IsZero()` using only one type doesn't seem sufficient given the number of allowed inputs. The test coverage was pretty good and we did find a few issues already.

Reply:
> What do you propose I call a test case structure that tests outer with zero value default, inner struct block, optional with a default value? I could just call them TestStructN and keep more descriptive test names. I'm not sure it is going to help though.

If I understand the scenario you just laid out: NonPropagatingBlock? Because it doesn't propagate the defaults from the inner struct.

Reply:
I have re-written tests and now I believe they are more readable. I still opted to cover some `struct` vs. `pointer` cases separately even though they may hit the same code path - otherwise it feels it's too white-box approach and I'm overfitting the tests to implementation.

Reply:
The tests are _much_ better now, thank you!